docs/strings.md

```diff
@@ -0,0 +1,464 @@
+
+Draft proposal for a core string library
+----------------------------------------
+
+The Verona standard string should define a string interface and a small set of concrete instantiations of this interface.
+The core string class should implement the minimum required for generic implementation of string operations, which can be implemented either using dynamic dispatch over the string interface or as generic reifications over the concrete types.
+We should make it possible to support a rich Unicode library that uses the same string objects as the standard library (and third-party libraries that use the same interface) and provides functions for iteration with unicode graphemes, locale-aware collation and other more complex operations that not all programs need.
+
+The interface for strings would look roughly like this:
+
+```verona
+/**
+ * String contains only 7-bit ASCII characters and is stored with each code
+ * point in an 8-bit integer.
+ */
+class StringEncodingASCII
+{
+	typedef CharType = U8;
+}
+/**
+ * String is stored as UTF-8-encoded text.
+ */
+class StringEncodingUTF8
+{
+	typedef CharType = U8;
+}
+/**
+ * String is stored as UTF-16-encoded text.
+ */
+class StringEncodingUTF16
+{
+	typedef CharType = U16;
+}
+/**
+ * String is stored as UTF-32-encoded text.
+ */
+class StringEncodingUTF32
+{
+	typedef CharType = U32;
+}
+/**
+ * Interface for other string encodings.  Any other type can be used as a valid
+ * string encoding but code that does not have any special cases for that
+ * encoding will be able to treat is as unknown.
+ */
+interface StringEncodingCustom
+{
+	typedef CharType;
+}
+
+/**
+ * The set of Unicode encodings.  All standard library functions that have
+ * special cases for specific encodings should handle these.
+ */
+typedef UnicodeStringEncoding = StringEncodingASCII |
+                                StringEncodingUTF8 |
+                                StringEncodingUTF16 |
+                                StringEncodingUTF32;
+/**
+ * The full set of possible string encodings, can be extended by third-party code.
+ */
+typedef StringEncoding = StringEncodingASCII |
+                         StringEncodingUTF8 |
+                         StringEncodingUTF16 |
+                         StringEncodingUTF32 |
+                         StringEncodingCustom;
+
+/**
+ * Character type, stores a unicode code point.
+ */
+typedef Unichar = U32;
+
+// Placeholder until we have syntax for value types.
+value Range
+{
+	U64 index;
+	U64 length;
+}
+
+/**
+ * Interface for string types.  Defines only the primitive operations that all
+ * strings must implement.
+ */
+interface String
+{
+	/**
+	 * Return the length of the string in Unicode code points.
+	 */
+	func length(self: readonly & String) : size_t;
+
+	/**
+	 * The encoding that this string can most easily convert to.  This will
```

Comment:
"most easily convert" makes it sound like this is a property of the pair of the string and its current encoding?

Reply:
Yes, assuming that a string is internally represented as some encoding.